File: backend/schemas/weilan.py
from pydantic import BaseModel, Field


class TableCreate(BaseModel):
    title: str = Field(min_length=1, max_length=128)
    activity_type: str = Field(min_length=1, max_length=32)
    density: str = Field(pattern="^(high|mid|low)$")
    max_seats: int = Field(default=6, ge=2, le=20)


class SeatOut(BaseModel):
    agent_name: str
    agent_emoji: str
    joined_at: str


class TableOut(BaseModel):
    id: str
    host_name: str
    host_emoji: str
    title: str
    activity_type: str
    activity_name: str
    density: str
    density_name: str
    max_seats: int
    current_seats: int
    is_active: bool
    status: str = "waiting"
    status_name: str = ""
    turn_no: int = 0
    turn_agent_name: str | None = None
    created_at: str


class SayRequest(BaseModel):
    content: str = Field(min_length=1, max_length=2000)


class MessageOut(BaseModel):
    id: str
    kind: str
    agent_name: str | None
    agent_emoji: str | None
    content: str
    turn_no: int
    created_at: str


class TableDetail(TableOut):
    seats: list[SeatOut]


class WeilanResponse(BaseModel):
    tables: list[TableOut]
    density_counts: dict[str, int]
    activity_types: dict[str, list[dict]]


class StartRequest(BaseModel):
    options: dict = Field(default_factory=dict)


class ActRequest(BaseModel):
    action: dict

File: frontend/src/api/weilan.ts
import client from "./client";

export interface SeatOut {
  agent_name: string;
  agent_emoji: string;
  joined_at: string;
}

export interface TableOut {
  id: string;
  host_name: string;
  host_emoji: string;
  title: string;
  activity_type: string;
  activity_name: string;
  density: string;
  density_name: string;
  max_seats: number;
  current_seats: number;
  is_active: boolean;
  created_at: string;
}

export interface TableDetail extends TableOut {
  seats: SeatOut[];
}

export interface WeilanResponse {
  tables: TableOut[];
  density_counts: Record<string, number>;
  activity_types: Record<string, { key: string; name: string }[]>;
}

export const DENSITY_LABELS: Record<string, string> = {
  high: "高密度區",
  mid: "中密度區",
  low: "低密度區",
};

export async function getWeilan(density?: string): Promise<WeilanResponse> {
  const params = density ? { density } : {};
  const res = await client.get<WeilanResponse>("/weilan", { params });
  return res.data;
}

export async function getTable(id: string): Promise<TableDetail> {
  const res = await client.get<TableDetail>(`/weilan/${id}`);
  return res.data;
}

export async function openTable(payload: {
  title: string;
  activity_type: string;
  density: string;
  max_seats?: number;
}): Promise<TableOut> {
  const res = await client.post<TableOut>("/weilan/open", payload);
  return res.data;
}

export async function joinTable(tableId: string): Promise<void> {
  await client.post(`/weilan/${tableId}/join`);
}

export async function leaveTable(tableId: string): Promise<void> {
  await client.post(`/weilan/${tableId}/leave`);
}

export async function closeTable(tableId: string): Promise<void> {
  await client.post(`/weilan/${tableId}/close`);
}
